release-25.4: sql: fix top-level query stats when "inner" plans are present

yuzefovich opened the pull request:
Backport 1/1 commits from #155824.
Backport 2/2 commits from #156110.

/cc @cockroachdb/release

---

Previously, whenever we ran an "inner" plan (via `runPlanInsidePlan` helper), we ignored the top-level query stats for the "inner" plan. As a result, reads and writes performed by the inner plan weren't reflected in the "outer" top-level query stats. This is now fixed by adjusting the routines, apply joins, and recursive CTEs to propagate their metrics as ProducerMetadata objects.

Note that for routines the access to the DistSQL infra is rather difficult, so we plumbed the access via the planner straight into the DistSQLReceiver, and even though it's ugly, it should work in practice. The only alternative I see is adding the necessary reference into the `eval.Context`, but then it gets tricky to actually set that and ensure the right copy of the eval context is observed by all routines (plus we'd need to make the copy or restore the original state somehow), so I chose to not pursue it.

Epic: None

Release note (bug fix): Previously, CockroachDB didn't include reads and writes performed by routines (user-defined functions and stored procedures) as well as apply joins into `bytes read`, `rows read`, and `rows written` statement execution statistics, and this is now fixed. The bug has been present since before 23.2 version.

Release justification: bug fix.

yuzefovich committed:
sql: fix top-level query stats when "inner" plans are present

Previously, whenever we ran an "inner" plan (via `runPlanInsidePlan`
helper), we ignored the top-level query stats for the "inner" plan. As
a result, reads and writes performed by the inner plan weren't reflected
in the "outer" top-level query stats. This is now fixed by adjusting the
routines, apply joins, and recursive CTEs to propagate their metrics as
ProducerMetadata objects.

Note that for routines the access to the DistSQL infra is rather
difficult, so we plumbed the access via the planner straight into the
DistSQLReceiver, and even though it's ugly, it should work in practice.
The only alternative I see is adding the necessary reference into the
`eval.Context`, but then it gets tricky to actually set that and ensure
the right copy of the eval context is observed by all routines (plus
we'd need to make the copy or restore the original state somehow), so
I chose to not pursue it.

Release note (bug fix): Previously, CockroachDB didn't include reads and
writes performed by routines (user-defined functions and stored
procedures) as well as apply joins into `bytes read`, `rows read`, and
`rows written` statement execution statistics, and this is now fixed.
The bug has been present since before 23.2 version.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/155945)
<!-- Reviewable:end -->

yuzefovich commented:
Note that the conflicts (due to the refactor in #145934) were minor.

yuzefovich committed:
sql: unconditionally check DistSQL supportability

This commit hardens c17591e82ed02124aa138ca10d8e3491af7552ff. In
particular, in that change we disabled the usage of the Streamer
whenever some part of the plan isn't distributable. However, population
of `distSQLProhibitedErr` was left on a best-effort basis. In
particular, if we have `distsql=off` (as well as a couple of other
conditions), we'd skip the supportability check. This meant that we
could still end up using the Streamer in some illegal cases - one
example that I came up with is if we have a routine (which currently
prohibits distsql), it might access the RootTxn concurrently with the
LeafTxn access by the Streamer, which is no bueno. This commit makes it
so that we do DistSQL supportability check unconditionally. Note that
this shouldn't really have a performance impact - after all, we do
expect this check to be done pretty much all the time (unless someone
runs with `distsql=off`).

Additionally, this change happens to fix a nil pointer error around the
recent fix to top-level query stats in presence of routines. Namely,
that patch only set the metadata forwarder for local plans if we end up
using the RootTxn, but in a query where we happened to use the streamer
with routines, (before this patch) we'd end up with LeafTxn and unset
metadata forwarder. Now we'll have streamer disabled, so we'll have the
RootTxn and the forwarder will be set.

Release note: None

yuzefovich committed:
sql: harden recent fix to top-level query stats with routines

This commit relaxes the requirements for when it's safe to set the
routine metadata forwarder. Previously, we only set it when we end up
using the RootTxn in a local plan, but now we examine all possible kinds
of concurrency within the local plan to see whether it's actually safe
to set the metadata forwarder. This seems like a nice cleanup as well.

Note that this commit independently fixes the issue mentioned in the
previous commit.

Release note: None

michae2 reviewed (approved):
Looks like this is ready? <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @michae2 reviewed 8 of 10 files at r1, 2 of 2 files at r2, 6 of 6 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155945#-:-Ocm5bWfC_Ubh6dORTpV:bf7bfbw)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Yeah, I think so. TFTR!
